Add examples, integration tests and docs for simulated device and reader interaction

radumarias opened the pull request:
Simulated device and reader interaction:

```mermaid
sequenceDiagram
    autonumber
    Note over Device: Initialize session
    Device ->> Device: Create QR Code Engagement
    Device -) + Reader: Send QR Code
    Reader ->> - Reader: Establish Session
    Reader -) + Device: Request age_over_21
    Device -)- Reader: Send age_over_21
    Reader ->> Reader: Process age_over_21
    Note over Device, Reader: Session finished
```

Device perspective:

```mermaid
stateDiagram
    state Device {
        [*] --> SessionManagerInit: initialise
        SessionManagerInit --> SessionManagerEngaged: qr_engagement
        SessionManagerEngaged --> SessionManager: process_session_establishment
    }

    state SessionManagerInit {
        [*] --> [*]
    }

    state SessionManagerEngaged {
        [*] --> [*]
    }

    state Reader {
        [*] --> [*]
    }

    state SessionManager {
        [*] --> AwaitingRequest
        AwaitingRequest --> Signing: prepare_response
        Signing --> Signing: get_next_signature_payload
        Signing --> ReadyToRespond: submit_next_signature
        ReadyToRespond --> AwaitingRequest: retrieve_response
        AwaitingRequest --> Signing: handle_request
    }

    User --> Device
    SessionManagerInit --> Reader: qr_engagement
    Reader --> SessionManagerEngaged: establish_session
    ReadyToRespond --> Reader: handle_response
```
Reader's perspective:

```mermaid
stateDiagram
    state Device {
        [*] --> [*]
    }

    state Reader {
        SessionManager --> SessionManager: handle_response
    }

    User --> Device
    Device --> Reader: qr_engagement
    Reader --> Device: establish_session
    Device --> Reader
    Reader --> Device: new_request
```

**Integration tests:**
- add simulated device and reader interaction
- also for simulated device and reader interaction, add a more structured one using `state pattern`, `Mutex` and `Arc`
- one test from the holder's POV with a simulated reader, and vice-versa for another test

**Docs:**
- add example from the holder's POV with a simulated reader, and vice-versa for another example
- add diagram for simulated device and reader interaction
- add diagrams for simulated device perspective and simulated reader perspective
- add diagrams in readme and links to examples
- add ASCII diagrams in `lib.rs` and include examples from device and reader's perspective
- fix some docs

**Code:**
- add `allow invisible_characters` for clippy in a file

**CI:**
- upgrade `actions/checkout`
- add `--all-targets` to build so that it builds tests too (to make sure they don't fall behind)
- add `--all-targets` to `clippy` so that it runs on tests too (it will run o tests too but I assume it's fine)
    - don't explicitly run in in `macros` because `--all-targets` takes care of that
- add `--all` to `fmt --check` so that it runs on examples too
    - don't explicitly run in in `macros` because `--all` takes care of that

radumarias committed:
Add examples member and an example for connection and data exchange between a device and a reader

radumarias committed:
add 'tags' feature to serde_cbor

radumarias committed:
fix path in CI for examples

radumarias committed:
organize example and added more docs

wyc commented on examples/README.md:
```diff
@@ -0,0 +1,31 @@
+# isomdl exmples
+
+# Connection and data exchange between a device and a reader
```
Would prefer the use of "simulated device and reader" throughout this example and its docs because there is not actually a device nor reader, and there's no QR/BLE exchange. Only DeviceEngagement and in-memory channels.

radumarias replied:
sure, changed

radumarias committed:
use 'simulated device and reader' terminology

radumarias committed:
- add example in readme
- add script to replace examples code in readme and lib.rs with code from actual file, we use a hidden INCLUDE specific tag to specify what file to include
  - it will replace the code in markdown code block with the code from file, keeping the import marker
  - when run next time it will update the code content from the file
- fix some docs
- add `invisible_characters` for clippy in a file

CI:
- upgrade actions/checkout
- add `--all-targets` to build so that it builds examples too
- run examples along with tests
- add ``--all-targets` to clippy so that it runs on examples too
- add `--all` to fmt so that it runs on examples too

radumarias committed:
removed unused text

radumarias committed:
add integration tests for a simulated device and reader interaction
in integration tests we use common code from examples and wrap the call

radumarias committed:
sign device response
restructure example
copied examples data to dedicated `examples_data` folder

radumarias committed:
rename integration test

radumarias committed:
run `sync-doc-includes` for all `.rs` files

radumarias committed:
don't rim spaces from lines when processing INCLUDE marker

radumarias committed:
split the example in two, one as before for a more structured version and another one for basic approach (will use this in docs as it's shorter)

radumarias committed:
added mermaid diagram in `readme` for simulated device and reader interaction

radumarias commented:
![Screenshot from 2024-07-08 11-27-20](https://github.com/radumarias/isomdl/assets/8073584/a35887bb-b182-4e0f-9669-f7f979f141d1)

radumarias committed:
updated diagram for simulated device and reader interaction

radumarias committed:
updated diagram for simulated device and reader interaction

radumarias committed:
updated code in readme

radumarias committed:
use full `Library` terminology in readme

radumarias committed:
move `example_data` to `example/data`
move ``*.sh` to `scripts`
added readme for scripts

radumarias committed:
separate simulated device and reader interaction in 2 examples from each side perspective
use these in readme and `lib.rs`

radumarias committed:
run `fmt` just one time but with `--all` which will run it in all members
run `clippy` just one time but with `--all-targets` which will run it in all members
add start_path as mandatory arg to sync-rustdoc-includes.sh
add mandatory arg to sync-md-includes.sh to specify path to md file
run scripts tests in ci

radumarias committed:
remove unnecessary `cd ..` when running examples, as we moved to a different step and shell restarts at $HOME path

radumarias committed:
add a special marker `#[spruce_example(run(skip=true))]` for examples that we should skip run in CI
this is added to examples which are run also from integration tests wrappers
remove ` --all-targets` from `cargo test` as with that present it doesn't run rustdocs

radumarias committed:
rename the example ignore marker to `#[spruce_example(ignore)]`
print with colors when running exmples

radumarias committed:
ignore running common module from examples

radumarias committed:
add `readme` for examples

radumarias committed:
Update lib.rs

radumarias committed:
Update README.md

radumarias committed:
fix rustdocs

radumarias committed:
Update README.md

radumarias committed:
Update lib.rs

radumarias committed:
Update README.md

radumarias committed:
fix rustdocs fmt

radumarias committed:
Update lib.rs

radumarias committed:
Update lib.rs

radumarias committed:
Update README.md

radumarias committed:
Update lib.rs

radumarias committed:
Update README.md

radumarias committed:
Update lib.rs

radumarias committed:
add diagrams for flow in examples readme

radumarias committed:
add `markdown` to block code in scripts readme

radumarias committed:
document `handle_request` transition and internal states for `device::SessionManager`

radumarias committed:
document `handle_request` transition and internal states for `device::SessionManager`

radumarias committed:
add links in readme for `on_simulated_device` and `on_simulated_reader`

radumarias committed:
rename `simulated_device_and_reader_*` examples

radumarias committed:
extent `INCLUDE` marker to inner comments also, like `///```
add test for this case also
add links to examples in readme

radumarias committed:
fix CI scripts tests

radumarias committed:
fix CI scripts tests

radumarias committed:
fix CI scripts tests

radumarias committed:
fix CI scripts tests

radumarias committed:
move scripts tests from CI to scripts folder
create run examples script and call that from CI

radumarias committed:
fix cd to parent and back with `cd -`

radumarias committed:
move some runs to one line in CI

sbihel commented:
Sorry for the delay, I still need to properly review your PR, but some early comments:
- I'm not a fan of having scripts to move code about. If you would like to include the README in the root documentation of the crate, I would prefer a `#![doc = include_str!("../README.md")]`. Obviously it includes some restrictions (e.g. rustdoc doesn't support mermaid to diagrams need to be converted to ASCII diagrams, the README currently presents both the library and the CLI but IMO it's fine to focus on the library and simply add a link to the CLI's README) but IMO it's acceptable.
- If you want to execute examples in the CI, I would prefer to run `cargo test --examples` as part of the CI (it does mean that then some `#[test]` need to be added to the examples which means they're a bit less straightforward to follow but IMO it's acceptable)

radumarias commented:
thank you, I'm changing these

On Wed, Jul 17, 2024 at 4:44 PM Simon Bihel ***@***.***>
wrote:

> Sorry for the delay, I still need to properly review your PR, but some
> early comments:
>
>    - I'm not a fan of having scripts to move code about. If you would
>    like to include the README in the root documentation of the crate, I would
>    prefer a #![doc = include_str!("../README.md")]. Obviously it includes
>    some restrictions (e.g. rustdoc doesn't support mermaid to diagrams need to
>    be converted to ASCII diagrams, the README currently presents both the
>    library and the CLI but IMO it's fine to focus on the library and simply
>    add a link to the CLI's README) but IMO it's acceptable.
>    - If you want to execute examples in the CI, I would prefer to run cargo
>    test --examples as part of the CI (it does mean that then some #[test]
>    need to be added to the examples which means they're a bit less
>    straightforward to follow but IMO it's acceptable)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/radumarias/isomdl/pull/1#issuecomment-2233364911>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5TC4FFTWK2ELDNPIY3RLLZMZYMTAVCNFSM6AAAAABKONJHTKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEMZTGM3DIOJRGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
And in the end, it's not the years in your life that count. It's the life
in your years.

radumarias committed:
Run examples from integration tests

- Create two integration tests as wrappers over `on_simulated_device`
and `on_simulated_reader`.
Like this, we run examples along with other tests.
- Remove the run examples script.

radumarias commented:
@sbihel
>If you want to execute examples in the CI, I would prefer to run cargo test --examples as part of the CI (it does mean that then some #[test] need to be added to the examples which means they're a bit less straightforward to follow but IMO it's acceptable)

Added integration tests as wrappers that run two more examples, like for the first 2 examples. Didn't found an elegant way to mark them with #[test] and run directly from `examples`. Let me know if you think this is not ok.

radumarias committed:
Remove example code from README.md and restructure diagrams.

- Remove example code from README.md and just leave the links to example files.
- Restructure diagrams to indicate clearer the internal states.
- In `lib.rs` we use `asciiflow` diagrams.
Also, we import code with `#![doc = include_str!`

radumarias commented:
@sbihel 
>I'm not a fan of having scripts to move code about. If you would like to include the README in the root documentation of the crate, I would prefer a #![doc = include_str!("../README.md")]. Obviously it includes some restrictions (e.g. rustdoc doesn't support mermaid to diagrams need to be converted to ASCII diagrams, the README currently presents both the library and the CLI but IMO it's fine to focus on the library and simply add a link to the CLI's README) but IMO it's acceptable.

I created `asciiflow` diagrams and use those in `lib.rs`. Also we import examples with `#![doc = include_str!`. Like this the code and readme are separated.

In readme I changed a bit the diagrams to be more clearer of the inner states in `SessionManager`. Also removed examples and use links to them (as here links to examples works fine, curious it they work the same on `crates.io`).

radumarias commented:
@sbihel @cobward what is the correct flow to sign the `CoseSign1`? I'm now using a key loaded from resources, as I've seen it's done in `wallet-sdk-rs` also, but shouldn't we actually use the session key and verify the same? If not then from where the reader will take the key to verify?
Also didn't found how the reader verify the signature. In `wallet-sdk-rs` I only found `sign`, not `verify`.

I'm loading key here https://github.com/radumarias/isomdl/pull/1/files#diff-6c776869f59bd92a929192880fb0cec33589dae3f02f1c90e5f1b84f86ddb4c9R119

cobward commented:
For clarity there are 2 CoseSign1's:
1. The `IssuerAuth` which is the MobileSecurityObject (MSO) signed by the credential issuer.
2. The `DeviceSignature` which signs over the session data and the disclosed data.

Presumable you're talking about the `DeviceSignature` here, but we also need to talk about the MSO. The MSO contains the public key of a keypair that is stored securely on the device where the mDL is held. Since the MSO is signed by the credential issuer, this public key is "endorsed" as an entity that can be "trusted" to present this mDL. The `DeviceSignature` is created by this keypair, by signing over the session data (including the public session keys).

> but shouldn't we actually use the session key and verify the same?

If we only used the session key, it would only bind the transaction to this session, not binding it back to the issuer.

> If not then from where the reader will take the key to verify?

The reader takes the public key from the MSO.

I hope that answers your questions?

cobward commented:
It also would be worth looking here to see how we generate an mDL in the existing tests:
https://github.com/radumarias/isomdl/blob/main/src/issuance/mdoc.rs#L623-L638

cobward commented:
> Added integration tests as wrappers that run two more examples, like for the first 2 examples. Didn't found an elegant way to mark them with #[test] and run directly from examples. Let me know if you think this is not ok.

I actually think we should just move the `examples` to `tests`. I've seen this plenty of times across the Rust ecosystem where the integration tests are the only examples in the repo.

radumarias commented:
@cobward thank you for your answers, the reasoning of using issuer key also makes sense, buit
>The DeviceSignature is created by this keypair, by signing over the session data (including the public session keys).

Do we have some clear example how `DeviceSignature` is created and verified by combining issuer pub key with session key? They're concatenated (issuer pub + device session) and using KDF a new private key (from which we have pub also for `CoseSign`) or symmetric key (for `CoseMac0`) are generated on device and reader and used to sign and verify?

radumarias committed:
Move examples to integration tests. They can be used as examples too.

cobward commented:
The payload to be signed is obtained [here](https://github.com/radumarias/isomdl/blob/main/src/presentation/device.rs#L360-L364) and the signature is then submitted [here](https://github.com/radumarias/isomdl/blob/main/src/presentation/device.rs#L366-L377).

We don't have any examples in this repo of actually signing that payload, but you are basically doing the right thing here:
https://github.com/radumarias/isomdl/blob/bf7de0c4834b76ad17e45824b6d7a7d5765d2457/tests/common.rs#L99-L101

**But the signature should not be DER encoded.** See the relevant passage from RFC 8152 here:
>   The signature algorithm results in a pair of integers (R, S).  These
   integers will be the same length as the length of the key used for
   the signature process.  The signature is encoded by converting the
   integers into byte strings of the same length as the key size.  The
   length is rounded up to the nearest byte and is left padded with zero
   bits to get to the correct length.  The two integers are then
   concatenated together to form a byte string that is the resulting
   signature.

So rather than `signature.to_der().to_vec()` you should just use `signature.to_vec()`.

radumarias commented:
@cobward just not clear how to combine the key with issuer pub, should I concatenate with decide session key and do KDF to get a private from which I derive the pub?

cobward commented:
You don't need to combine the keys, no KDF is needed. KDF is only used in the generation of session keys, which is handled by this library and consumers of this library don't need to do anything.

radumarias commented:
@cobward 
> You don't need to combine the keys, no KDF is needed. KDF is only used in the generation of session keys, which is handled by this library and consumers of this library don't need to do anything.

was asking in the context of this https://github.com/radumarias/isomdl/pull/1#issuecomment-2249871079
>The MSO contains the public key of a keypair that is stored securely on the device where the mDL is held. Since the MSO is signed by the credential issuer, this public key is "endorsed" as an entity that can be "trusted" to present this mDL. The DeviceSignature is created by this keypair, by signing over the session data (including the public session keys).

which is the private key for signing? as the MSO contains only the pub I assume

cobward commented:
In case it wasn't clear I can summarize the issuance process:

1. The mobile device generates the "device keypair".
2. The mobile device sends the public key of the "device keypair" to the issuer.
3. The issuer creates the mdoc, with the public key of the "device keypair" inside the MSO.
4. The issuer sends the mdoc to the mobile device.

Then when the mobile device needs to present the mdoc:

1. The mobile device initiates the transaction with the reader, creating a secure session with session keys.
2. The reader sends a request to the mobile device.
3. The mobile device constructs a response. Part of this response is a signature from "device keypair" to prove ownership of the public key that is in the mdoc.
4. The verifier receives the response:
  a. It verifies the `DeviceSignature` using the public key from the MSO.
  b. It verifies the `IssuerAuth` using the public key from the x509 certificate in the IssuerAuth header.
  c. It verifies the trust chain of this x509 certificate.
  
  
Does this clarify it for you?

radumarias commented:
>The issuer creates the mdoc, with the public key of the "device keypair" inside the MSO.

thank you very much for the detailed explanation, this part I was missing, I assumed the MSO includes the issuer pub

So in this case I assume the device holds its private alongside with Mdoc(s) then.


btw, is there a wallet app that uses `isomdl`? I would be interested to see a real world working example

Or `wallet-sdk-rs` with other repos for kotlin does something similar? Or is that a completely different story?
I looked at this https://github.com/spruceid/wallet-sdk-rs but the Kotlin code is not actually doing anything.

cobward commented:
> So in this case I assume the device holds its private alongside with Mdoc(s) then.

Yep!

> btw, is there a wallet app that uses isomdl?

We have an example app [here](https://github.com/spruceid/wallet-sdk-kt/tree/main/example) (which uses isomdl through wallet-sdk-rs).

radumarias commented:
>So rather than signature.to_der().to_vec() you should just use signature.to_vec().

Right, don't even know from where I've got the `to_der()` :) fixing it right now.

radumarias committed:
Remove `to_der()` when setting signature.

radumarias committed:
Remove `to_der()` when setting signature.

radumarias committed:
Resolve small typo

cobward commented on scripts/README.md:
Are we using these `INCLUDE` files anywhere now? Maybe we can remove the scripts directory entirely and remove it from the CI too?

radumarias replied:
removed those scripts

cobward commented on README.md:
```diff
@@ -10,7 +13,128 @@ This crate contains a CLI tool. Run the `--help` command to see what actions you
+Here are some examples of how to use the library. You can see more in [examples](examples) folder and read about in the
+dedicated [README](tests/README.md).
+
+### Eamples
```
```suggestion
### Examples
```

cobward commented on README.md:
```diff
@@ -10,7 +13,128 @@ This crate contains a CLI tool. Run the `--help` command to see what actions you
+
+## Library
+
+Here are some examples of how to use the library. You can see more in [examples](examples) folder and read about in the
```
```suggestion
Here are some examples of how to use the library. You can see more in [tests](tests) folder and read about in the
```

radumarias committed:
change the ISO mDL link

radumarias committed:
Remove scripts to resolve include links. Fix small typo.

cobward commented:
My last comments before I think we can look to merge this:

* The functions in `tests/common.rs` aren't actually used in the integration tests. Either remove common.rs, or use those functions and remove the dead_code allows.

* There is a lot of overlap between all four of the integration test files. Do you think there is a need for all four or could we just have a single well documented test?

radumarias commented:
> The functions in tests/common.rs aren't actually used in the integration tests. Either remove common.rs, or use those functions and remove the dead_code allows.

They are indeed used in `on_simulated_device` and `on_simulate_reader` but for some reason the compiler sees them as unused. Don't know why, maybe because the `common` module is defined 2 times, once in each file? But I didn't found other way to use it in both. Maybe use `include!` macro?

> There is a lot of overlap between all four of the integration test files. Do you think there is a need for all four or could we just have a single well documented test?

The separation was made also based on this feedback https://sprucesystems.slack.com/archives/C07AS9P1A75/p1720450066785869 and I also think it's easier for users to follow several scoped examples than the whole flow of the interaction.

Let me know if I should change anything.

radumarias committed:
Fix some better grammar.
